# EDA - Qualidade da Água | Projeto AquaSense

## Contexto
Este notebook apresenta a Análise Exploratória de Dados (EDA) de um dataset de qualidade da água, utilizado no contexto do projeto AquaSense.

O conjunto de dados reúne medições de parâmetros físico-químicos da água em diferentes países (Canadá, China, Irlanda, e Estados Unidos), incluindo também o índice CCME WQI.

## Objetivo desta etapa
O objetivo desta análise é compreender o comportamente dos dados antes de qualquer modelagem, buscando identificar padões, incosistências e limitações do dataset.

De forma mais específica, pretende-se:
- Entender a estrutura e composição dos dados
- Identificart problemas de qualidade (valores nulos, inconsistências, etc.)
- Analisar distribuições e possíveis outliers
- Explorar relações entre variáveis
- Identificar parâmetros compatíveis com a Resolução CONAMA 357/2005

Essa estapa é fundamental para orientar decisões futuras, especialmente na construção de um rótulo classificatório adaptado ao contexto brasileiro.

## Ferramentas de visualização
Para análise gráfica, será utilizada a biblioteca Plotly, devido à sua capacidade de gerar visualizações interativas, facilitando a interpretação dos dados.

A análise será conduzida de forma progressiva, partindo de uma visão geral dos dados até uma abordagem orientada ao objetivo do projeto.

# Preparação do Ambiente
Nesta seção, são importadas as bibliotecas necessárias para manipulação, análise e visualização dos dados. As ferramentas escolhidas permitem trabalhar com estruturas tabulares, realizar operações numéricas e construir gráficos que auxiliam na interpretação dos padrões do dataset.

In [2]:
# Manipulação de dados
import pandas as pd
import numpy as np

# Visualização de dados
import plotly.express as px
import plotly.graph_objects as go

In [3]:
import pandas as pd

# Carregar o dataset local
df = pd.read_csv("../../dataset/Combined_dataset.csv")

df.head()

,Country,Area,Waterbody Type,Date,Ammonia (mg/l),Biochemical Oxygen Demand (mg/l),Dissolved Oxygen (mg/l),Orthophosphate (mg/l),pH (ph units),Temperature (cel),Nitrogen (mg/l),Nitrate (mg/l),CCME_Values,CCME_WQI
0,Canada,SE649035-145565,River,12-01-1974,0.059248,1.30,8.1500,0.011917,8.07500,9.885,0.343917,11.73155,100.0,Excellent
1,Canada,SE649035-145565,River,12-01-1975,0.039821,1.38,7.8000,0.009417,7.73333,10.150,0.449083,11.82009,100.0,Excellent
2,Canada,SE649035-145565,River,12-01-1976,0.031341,2.23,7.8000,0.011000,7.46667,10.235,0.220750,14.87472,100.0,Excellent
3,Canada,SE649035-145565,River,12-01-1977,0.020501,1.61,8.1500,0.012333,7.78333,11.116,0.572250,15.89293,100.0,Excellent
4,Canada,SE649035-145565,River,12-01-1978,0.020023,1.64,4.3708,0.006182,7.10000,7.068,0.371091,15.22888,100.0,Excellent


# Conhecimento Estrutural do Dataset

## Dimensão do Dataset
Antes de analisar qualquer conteúdo, precisamos entender o tamanho da base. Isos porque a escala dos dados influencia diretamente toda a análise.


---



**Pergunta que queremos responder**


*   Quantas observações existem?
*   Quantas variáveis estão disponíveis?


---


**Objetivo**


*   Identificar o volume dos dados
*   Ter uma noção da dimensão do problema






In [ ]:
df.shape

(2827977, 14)



*   2.827.977 - número de linhas (registros)
*   14 - números de colunas (variáveis)

Essa etapa é importante porque nos ajuda a responder a seguinte pergunta: qual é o tamanho da base com a qual estamos trabalhando? Saber isso desde o início permite que a análise seja conduzida com mais consciência, já que o comportamento dos dados e o tipo de cuidado analítico necessário podem mudar bastante dependendo da escala da base.





## Identificação das variáveis
Agora que sabemos o tamanho da base, precisamos entender o que ela contém.


---


**Pergunta que queremos responder**


*   Quais informações estão disponíveis no dataset?


---
**Objetivo**


*   Mapear todas as variáveis
*   Identificar possíveis variáveis importantes








In [ ]:
colunas_df = pd.DataFrame({
    'Variáveis': df.columns

})

display(colunas_df)

,Variáveis
0,Country
1,Area
2,Waterbody Type
3,Date
4,Ammonia (mg/l)
5,Biochemical Oxygen Demand (mg/l)
6,Dissolved Oxygen (mg/l)
7,Orthophosphate (mg/l)
8,pH (ph units)
9,Temperature (cel)


A partir da organização das colunas em formato de tabela, é possível identificar diferentes tipos de variáveis presentes no dataset, cada uma com um papel específico dentro da análise.

De forma geral, as variáveis podem ser classificadas em três grupos principais:

* **Variáveis de contexto:** incluem informações como Country, Area, Waterbody Type e Date. Essas variáveis ajudam a entender onde e quando os dados foram coletados, sendo importantes para análises comparativas e segmentações, mas não representam diretamente a qualidade da água.

* **Variáveis físico-químicas (potenciais variáveis de entrada):** incluem parâmetros como Ammonia, Biochemical Oxygen Demand, Dissolved Oxygen, Orthophosphate, pH, Temperature, Nitrogen e Nitrate. Essas variáveis são diretamente relacionadas às características da água e possuem grande relevância para a análise, pois são os principais fatores que influenciam sua qualidade.

* **Variáveis de resultado (potenciais variáveis-alvo):** destacam-se CCME_Values e CCME_WQI. A variável CCME_Values representa um índice numérico de qualidade da água, enquanto CCME_WQI apresenta uma classificação categórica (como "Excellent"). Essas variáveis são fortes candidatas a variável-alvo em modelos de machine learning, pois sintetizam a qualidade da água a partir dos parâmetros analisados.


Além disso, observa-se que os nomes das colunas estão, em geral, bem definidos e coerentes com os parâmetros analisados. No entanto, alguns nomes mais longos, como Biochemical Oxygen Demand, podem exigir padronização futura para facilitar a manipulação dos dados.

## Visualização inicial
Agora vamos olhar os dados como eles realmente aparecem.


---


**Pergunta que queremos responder**


*   Como os dados estão organizados na prática?


---


**Objetivo**

*   Ver formato real dos dados
*   Identificar inconsistências visuais





In [ ]:
df.head()
df.tail()

,Country,Area,Waterbody Type,Date,Ammonia (mg/l),Biochemical Oxygen Demand (mg/l),Dissolved Oxygen (mg/l),Orthophosphate (mg/l),pH (ph units),Temperature (cel),Nitrogen (mg/l),Nitrate (mg/l),CCME_Values,CCME_WQI
2827972,Ireland,"Cork Harbour, Moy Killala, YELLOW (KNOCK)_020",River,21-06-2022,0.024,1.4,4.2930,0.025,7.9,16.2,0.37,0.37,100.0,Excellent
2827973,Ireland,"Cork Harbour, Moy Killala, YELLOW (KNOCK)_020",River,28-09-2022,0.038,1.4,4.4955,0.025,7.9,11.4,0.54,0.54,100.0,Excellent
2827974,Ireland,"Cork Harbour, Moy Killala, YELLOW (KNOCK)_020",River,29-11-2022,0.035,1.4,4.4640,0.016,7.6,6.2,0.80,0.79,100.0,Excellent
2827975,Ireland,"Cork Harbour, Moy Killala, YELLOW (KNOCK)_020",River,07-02-2023,0.046,1.4,4.1760,0.012,8.0,7.8,0.95,1.30,100.0,Excellent
2827976,Ireland,"Cork Harbour, Moy Killala, YELLOW (KNOCK)_020",River,11-04-2023,0.020,1.2,4.3245,0.025,7.9,7.9,0.49,1.30,100.0,Excellent


A visualização das primeiras e últimas linhas permite observar como os dados aparecem na prática.

Os registros seguem um padrão consistente, em que cada linha representa uma medição de qualidade da água associada a um local e uma data específica. Os valores numéricos aparentam estar dentro de faixas esperadas, indicando coerência inicial dos dados.

Também é possível notar a presença de muitas casas decimais em algumas variáveis e a necessidade de tratar a coluna de data para análises futuras.

De forma geral, não foram identificadas inconsistências visuais relevantes neste primeiro contato, permitindo avançar para as próximas etapas da análise.

## Estrutura técnica do dataset
Nesta etapa, utilizamos o método df.info() para visualizar a estrutura geral do dataset, incluindo quantidade de registros, número de colunas, tipos de dados e uso de memória.



---

**Pergunta que queremos responder**


*   Como os dados estão estruturados tecnicamente e quais pontos de atenção precisam ser observados antes de aprofundar a análise?



---

**Objetivo**





*   Compreender a organização interna do dataset
*   Verificar qualidade de entradas e tipos de variáveis


*   Verificar possíveis inconsistências de tipagem e aspectos que possam impactar as próximas etapas da EDA



In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2827977 entries, 0 to 2827976
Data columns (total 14 columns):
 #   Column                            Dtype  
---  ------                            -----  
 0   Country                           object 
 1   Area                              object 
 2   Waterbody Type                    object 
 3   Date                              object 
 4   Ammonia (mg/l)                    float64
 5   Biochemical Oxygen Demand (mg/l)  float64
 6   Dissolved Oxygen (mg/l)           float64
 7   Orthophosphate (mg/l)             float64
 8   pH (ph units)                     float64
 9   Temperature (cel)                 float64
 10  Nitrogen (mg/l)                   float64
 11  Nitrate (mg/l)                    float64
 12  CCME_Values                       float64
 13  CCME_WQI                          object 
dtypes: float64(9), object(5)
memory usage: 302.1+ MB


O dataset possui 2.827.977 entradas, o que indica uma base de dados muito rica e pesada (ocupa cerca de 302 MB de memória RAM).

As colunas de medições químicas (Amônia, Oxigênio, pH, Temperatura, etc.) estão corretamente tipadas como float64 (números decimais).

A coluna Date está como object (texto).
Ponto de atenção: Para fazer análises ao longo do tempo, precisaremos convertê-la para o formato de data real futuramente.

## Primeiras impressões sobre o dataset
Nesta etapa, reunimos as observações iniciais a partir da visualização da base, considerando sua dimensão, estrutura e aparência geral dos dados.



---

O dataset apresenta um volume elevado de dados, com 2.827.977 registros e 14 variáveis, indicando uma base robusta para exploração. A organização das colunas sugere uma divisão clara entre variáveis de contexto, parâmetros físico-químicos e indicadores de qualidade da água, o que favorece diferentes tipos de análise.

A visualização das primeiras e últimas linhas indica um padrão consistente nos registros, sem inconsistências visuais relevantes neste primeiro contato.

Por outro lado, já é possível identificar pontos de atenção, como a coluna Date em formato object, que exigirá conversão para análises temporais, além do tamanho da base, que pode impactar o processamento. De forma geral, os dados se mostram bem estruturados, permitindo avançar para as próximas etapas da EDA.

# Qualidade dos Dados


## Checagem de valores nulos
Nesta etapa, analisamos a presença de valores ausentes no dataset. A identificação de dados faltantes é fundamental, pois a ausência de informações pode comprometer análises estatísticas, visualizações e o desempenho de modelos de machine learning.



---

**Pergunta que queremos responder**


*   Existem valores nulos no dataset que possam impactar a análise?


---

**Objetivo**


*   Verificar a existência de valores ausentes nas colunas


*   Identificar possíveis variáveis com maior incidência de nulos
*   Avaliar a necessidade de tratamento dos dados antes das próximas etapas



In [5]:
df.isnull().sum()

,0
Country,0
Area,0
Waterbody Type,0
Date,0
Ammonia (mg/l),0
Biochemical Oxygen Demand (mg/l),0
Dissolved Oxygen (mg/l),0
Orthophosphate (mg/l),0
pH (ph units),0
Temperature (cel),0


In [6]:
percentual_nulos = (df.isnull().sum() / len(df)) * 100

pd.DataFrame({
    "Percentual de Nulos (%)": percentual_nulos
}).sort_values(by="Percentual de Nulos (%)", ascending=False)

,Percentual de Nulos (%)
Country,0.0
Area,0.0
Waterbody Type,0.0
Date,0.0
Ammonia (mg/l),0.0
Biochemical Oxygen Demand (mg/l),0.0
Dissolved Oxygen (mg/l),0.0
Orthophosphate (mg/l),0.0
pH (ph units),0.0
Temperature (cel),0.0


A análise indica que o dataset não possui valores nulos em nenhuma das colunas, demonstrando um alto nível de completude dos dados. Isso é um ponto positivo, pois elimina, neste momento, a necessidade de aplicação de técnicas de tratamento como imputação ou remoção de registros.

Essa característica sugere que o processo de coleta ou consolidação dos dados foi bem estruturado, permitindo que a análise prossiga diretamente para etapas mais avançadas, sem necessidade de intervenções relacionadas à ausência de informações.

## Análise de duplicatas

Nesta etapa, verificamos a existência de registros duplicados no dataset. A identificação de duplicatas é importante porque repetições indevidas podem distorcer análises estatísticas, frequências e resultados de modelos, comprometendo a confiabilidade da base.



---

**Pergunta que queremos responder**


*   Existem registros duplicados no dataset que possam comprometer a análise?


---

**Objetivo**


*   Identificar a presença de registros repetidos
*   Avaliar se há risco de distorção nos resultados


*   Verificar a necessidade de remoção ou tratamento dessas duplicatas







In [8]:
# Quantidade de duplicatas
duplicatas = df.duplicated().sum()
print(f"Quantidade de registros duplicados: {duplicatas}")

# Visualização das duplicatas (se existirem)
df[df.duplicated()]

Quantidade de registros duplicados: 9971


,Country,Area,Waterbody Type,Date,Ammonia (mg/l),Biochemical Oxygen Demand (mg/l),Dissolved Oxygen (mg/l),Orthophosphate (mg/l),pH (ph units),Temperature (cel),Nitrogen (mg/l),Nitrate (mg/l),CCME_Values,CCME_WQI
3516,Canada,SE616397-135001,River,10-04-2004,0.032844,2.13333,9.824000,0.0610,8.2300,12.0000,0.8900,5.836820,93.116725,Good
3522,Canada,ESGV09089001,River,12-01-2009,0.043792,2.12500,9.824000,0.0030,7.0800,26.6600,0.0880,2.001004,93.188249,Good
3524,Canada,HR21001,River,12-01-2012,0.139748,1.32750,11.400000,0.0040,7.0200,20.0500,0.4880,3.375587,100.000000,Excellent
3525,Canada,FI34398,Lake,02-12-2007,0.004508,2.13333,9.824000,0.0025,6.3729,2.2333,0.0570,5.836820,100.000000,Excellent
3562,Canada,FISW_32,Lake,10-04-2012,0.005796,0.80000,9.824000,0.0011,6.7234,6.6875,0.0038,0.545995,100.000000,Excellent
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2802991,Ireland,"Blackwater_Munster, Blackwater (Munster)",River,18-12-2012,0.038000,1.40000,4.972500,0.0250,7.8000,11.1500,1.4220,1.300000,100.000000,Excellent
2803180,Ireland,"Blackwater_Munster, Blackwater (Munster)",River,27-04-2013,0.038000,1.40000,4.972500,0.0250,7.8000,11.1500,1.4220,1.300000,100.000000,Excellent
2803748,Ireland,"Blackwater_Munster, Blackwater (Munster)",River,19-10-2011,0.038000,1.40000,4.972500,0.0250,7.8000,11.1500,1.4220,1.300000,100.000000,Excellent
2803768,Ireland,"Blackwater_Munster, Blackwater (Munster)",River,12-06-2014,0.038000,1.40000,11.684103,0.0250,7.8000,11.1500,1.4220,1.300000,100.000000,Excellent


## Correções básicas
Nesta etapa, são identificadas possíveis correções iniciais necessárias para melhorar a consistência e a adequação do dataset. Essas correções não representam um pré-processamento completo, mas sim ajustes básicos que facilitam as próximas etapas da análise.

In [9]:
df['Date'] = pd.to_datetime(df['Date'], format='%d-%m-%Y')

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2827977 entries, 0 to 2827976
Data columns (total 14 columns):
 #   Column                            Dtype         
---  ------                            -----         
 0   Country                           object        
 1   Area                              object        
 2   Waterbody Type                    object        
 3   Date                              datetime64[ns]
 4   Ammonia (mg/l)                    float64       
 5   Biochemical Oxygen Demand (mg/l)  float64       
 6   Dissolved Oxygen (mg/l)           float64       
 7   Orthophosphate (mg/l)             float64       
 8   pH (ph units)                     float64       
 9   Temperature (cel)                 float64       
 10  Nitrogen (mg/l)                   float64       
 11  Nitrate (mg/l)                    float64       
 12  CCME_Values                       float64       
 13  CCME_WQI                          object        
dtypes: datetime64[ns](

A partir das análises anteriores, foi possível identificar alguns ajustes iniciais necessários para melhorar a consistência do dataset.

Um dos principais pontos foi a coluna Date, que inicialmente estava em formato textual (object). Para permitir análises temporais mais adequadas, essa variável foi convertida para o tipo datetime, tornando possível a realização de estudos ao longo do tempo, como tendências e padrões sazonais.

Além disso, foram identificados 9.971 registros duplicados no dataset. Neste momento, optou-se por não realizar a remoção dessas duplicatas, mantendo a base original para preservar a integridade dos dados durante a análise exploratória.

De forma geral, essas correções iniciais contribuem para tornar a base mais adequada para as próximas etapas, sem alterar significativamente sua estrutura original.

# Estatísticas Descritivas

## Estatísticas descritivas iniciais com describe()
Nesta etapa, utilizamos o método describe() para obter um resumo estatístico das variáveis numéricas do dataset. Essa função apresenta medidas importantes como média, mediana, desvio padrão, valores mínimos e máximos e quartis, permitindo uma visão inicial mais completa sobre o comportamento dos dados.


---
**Pergunta que queremos responder**


*   Como as variáveis numéricas do dataset se comportam em termos de tendência central, dispersão e distribuição inicial?


---

**Objetivo**


*   Observar medidas de tendência central, como média e mediana
*   Avaliar a dispersão dos dados por meio do desvio padrão


*   Identificar a amplitude dos valores com base no mínimo e no máximo
*   Analisar os quartis para compreender melhor a distribuição das variáveis


*   Detectar indícios iniciais de assimetrias e possíveis valores extremos












In [11]:
df.describe()

,Date,Ammonia (mg/l),Biochemical Oxygen Demand (mg/l),Dissolved Oxygen (mg/l),Orthophosphate (mg/l),pH (ph units),Temperature (cel),Nitrogen (mg/l),Nitrate (mg/l),CCME_Values
count,2827977,2.827977e+06,2.827977e+06,2.827977e+06,2.827977e+06,2.827977e+06,2.827977e+06,2.827977e+06,2.827977e+06,2.827977e+06
mean,2007-05-15 06:45:57.940675328,1.171360e+00,4.886554e+00,1.000798e+01,7.104658e-01,7.735931e+00,1.183934e+01,5.209531e+00,4.766794e+00,8.504668e+01
min,1940-01-02 00:00:00,-5.000000e-03,-2.000000e+00,0.000000e+00,-4.000000e-03,-1.000000e+00,-5.260000e+00,0.000000e+00,0.000000e+00,3.130414e+01
25%,2003-07-03 00:00:00,3.000000e-02,1.600000e+00,9.860000e+00,4.000000e-02,7.550000e+00,8.990000e+00,7.800000e-01,1.173000e+00,7.715349e+01
50%,2008-10-23 00:00:00,5.500000e-02,2.700000e+00,1.020000e+01,1.070000e-01,7.780000e+00,1.146000e+01,4.000000e+00,4.500000e+00,9.059609e+01
75%,2014-12-09 00:00:00,3.170000e-01,2.830000e+00,1.100000e+01,2.270000e-01,8.000000e+00,1.420000e+01,6.320000e+00,4.940000e+00,1.000000e+02
max,2023-12-17 00:00:00,2.000000e+02,2.550000e+02,2.000000e+01,1.000000e+02,3.000000e+01,9.800000e+01,4.600000e+01,1.550000e+02,1.000000e+02
std,NaN,5.668927e+00,1.641363e+01,1.851052e+00,2.089436e+00,4.946871e-01,5.010197e+00,6.200423e+00,6.073781e+00,1.764665e+01


**Completude dos dados**

A estatística count indica que todas as variáveis possuem 2.827.977 registros, confirmando a ausência de valores nulos e a consistência da base para análise.


---


**Tendência central**

Ao observar as medidas de tendência central, especialmente média e mediana, percebe-se que algumas variáveis apresentam diferenças relevantes entre esses valores. Isso ocorre, por exemplo, em Ammonia, Orthophosphate, Biochemical Oxygen Demand, Nitrogen e Nitrate, cujas médias são superiores às medianas. Esse comportamento sugere distribuições assimétricas, nas quais a maior parte dos valores está concentrada em faixas mais baixas, enquanto alguns valores elevados puxam a média para cima. Esse resultado reforça a importância de não analisar apenas a média de forma isolada.


---


**Dispersão dos dados**

A análise do desvio padrão (std) mostra que algumas variáveis apresentam variabilidade considerável, como Biochemical Oxygen Demand, Nitrogen, Nitrate e Ammonia. Isso indica que os dados não estão uniformemente concentrados em torno da média, mas distribuídos de maneira mais espalhada. Já variáveis como pH e Dissolved Oxygen apresentam desvio padrão relativamente menor, sugerindo comportamento mais estável em comparação com as demais.


---


**Amplitude e valores extremos**

A comparação entre os valores mínimos e máximos revela amplitudes bastante elevadas em algumas variáveis, o que chama atenção para a presença de valores extremos. Além disso, a existência de valores mínimos negativos em variáveis como Ammonia, Biochemical Oxygen Demand, Orthophosphate, pH e Temperature merece atenção, pois alguns desses valores podem não fazer sentido físico ou podem estar associados a convenções laboratoriais, limites de detecção ou inconsistências de registro. Esses casos precisarão ser analisados com mais cuidado nas etapas seguintes.


---


**Quartis e indícios de outliers**

Os quartis ajudam a entender onde a maior parte dos dados está concentrada. Em várias variáveis, observa-se que o terceiro quartil (Q3) está muito distante do valor máximo, indicando a presença de valores muito acima do comportamento mais comum da base. Isso aparece com bastante força em variáveis como Ammonia, Biochemical Oxygen Demand, Orthophosphate e Nitrate. Esses resultados sugerem indícios iniciais de outliers, que ainda precisarão ser confirmados por análises visuais e métodos específicos, como boxplots e avaliação por intervalo interquartil.

---


**Cobertura temporal**

A variável Date indica registros entre 1940 e 2023, mostrando uma ampla cobertura temporal que pode ser explorada em análises futuras.


---

De forma geral, as estatísticas descritivas mostram que o dataset é amplo, completo e analiticamente rico, mas também revelam sinais importantes de assimetria, alta variabilidade em algumas variáveis e possíveis valores extremos. Esses achados são fundamentais porque orientam as próximas etapas da EDA: a análise da distribuição das variáveis, a visualização de outliers por histogramas e boxplots, e a avaliação mais crítica da consistência dos parâmetros em relação ao contexto do projeto.